### PR TITLE
recent fixes

### PR DESCRIPTION
**Pull request by majestrate**
enable router profiling by default and disable static link in release target

### CODE REVIEW

**Comment by majestrate** [comment]
also:
* reduce cpu usage on transit hops
* client side traffic replay filter for filtering replays from shit routers
* randomize intro selection on shift to even out load
* make clients lookup introsets from random not closest routers

**Comment by majestrate** [comment]
also:

* add profiling feedback from LSRM

File: llarp/path/path.hpp [hunk 1]
```diff
@@ -394,6 +403,8 @@ namespace llarp
       uint64_t m_ExitObtainTX            = 0;
       PathStatus _status;
       PathRole _role;
+      util::DecayingHashSet< TunnelNonce > m_UpstreamBloomFilter;
+      util::DecayingHashSet< TunnelNonce > m_DownstreamBloomFilter;
```
Comment (notlesh) [inline llarp/path/path.hpp]: Why are we calling these bloom filters if they're really just hash maps?

File: llarp/path/ihophandler.cpp [hunk 2]
```diff
@@ -34,4 +34,4 @@ namespace llarp
       return true;
     }
   }  // namespace path
-}  // namespace llarp
```
Comment (notlesh) [inline llarp/path/ihophandler.cpp]: Meaningless change, I usually catch these with `git commit -p`

File: Makefile [hunk 3]
```diff
@@ -160,7 +160,7 @@ debug-configure:
 
 release-configure: clean
 	mkdir -p '$(BUILD_ROOT)'
-	$(CONFIG_CMD) -DCMAKE_BUILD_TYPE=Release -DSTATIC_LINK=ON -DRELEASE_MOTTO="$(shell cat motto.txt)" -DCMAKE_C_FLAGS='$(CFLAGS)' -DCMAKE_CXX_FLAGS='$(CXXFLAGS)'
```
Comment (notlesh) [inline Makefile]: This would ideally be in its own PR

**Comment by majestrate** [comment]
On Sat, Jan 04, 2020 at 11:54:22AM -0800, Stephen Shelton wrote:
> notlesh requested changes on this pull request.
> 
> What's with the variable names including BloomFilter when they're not a bloom filter?
> 
> > @@ -394,6 +403,8 @@ namespace llarp
>        uint64_t m_ExitObtainTX            = 0;
>        PathStatus _status;
>        PathRole _role;
> +      util::DecayingHashSet< TunnelNonce > m_UpstreamBloomFilter;
> +      util::DecayingHashSet< TunnelNonce > m_DownstreamBloomFilter;
> 
> Why are we calling these bloom filters if they're really just hash maps?

they were decaying bloomfilters but then i changed them to be decaying hashsets, can remove the Bloom part if needed.

> 
> > @@ -34,4 +34,4 @@ namespace llarp
>        return true;
>      }
>    }  // namespace path
> -}  // namespace llarp
> 
> Meaningless change, I usually catch these with `git commit -p`
> 
> > @@ -160,7 +160,7 @@ debug-configure:
>  
>  release-configure: clean
>  	mkdir -p '$(BUILD_ROOT)'
> -	$(CONFIG_CMD) -DCMAKE_BUILD_TYPE=Release -DSTATIC_LINK=ON -DRELEASE_MOTTO="$(shell cat motto.txt)" -DCMAKE_C_FLAGS='$(CFLAGS)' -DCMAKE_CXX_FLAGS='$(CXXFLAGS)'
> 
> This would ideally be in its own PR

yeah maybe.

**Comment by jagerman** [comment]
This PR has quite a few completely unrelated things in it.

**Comment by majestrate** [comment]
On Sun, Jan 05, 2020 at 08:38:04PM -0800, Jason Rhinelander wrote:
> This PR has quite a few completely unrelated things in it.
yes it does

File: Makefile (hunk 3 above)
Comment (majestrate) [reply Makefile]: made #1022

File: llarp/path/path.hpp (hunk 1 above)
Comment (majestrate) [reply llarp/path/path.hpp]: updated variable names

File: llarp/router/outbound_session_maker.cpp [hunk 4]
```diff
@@ -121,21 +121,28 @@ namespace llarp
...
-    _nodedb->visit([&](const RouterContact &other) -> bool {
-      if(!_rcLookup->RemoteIsAllowed(other.pubkey))
-      {
-        return remainingDesired > 0;
-      }
-      if(randint() % 2 == 0
-         && !(_linkManager->HasSessionTo(other.pubkey)
-              || HavePendingSessionTo(other.pubkey)))
+    std::set< RouterID > exclude;
+    int itersLeft = 100;
```
Comment (notlesh) [inline llarp/router/outbound_session_maker.cpp]: The purpose and consequences of `itersLeft` is pretty unclear, and `100` might be best defined as a constant

Comment (majestrate) [reply llarp/router/outbound_session_maker.cpp]: it's to prevent infinite loops

**Comment by notlesh** [comment]
LGTM
